Expose dispatchedTo and dispatchedFrom annotations to the IPCTestingAPI

gavin-apple opened the pull request:
#### ead822965a61db398c199145facba856f51a3b7d
<pre>
Expose dispatchedTo and dispatchedFrom annotations to the IPCTestingAPI
<a href="https://bugs.webkit.org/show_bug.cgi?id=297269">https://bugs.webkit.org/show_bug.cgi?id=297269</a>
<a href="https://rdar.apple.com/problem/158126216">rdar://problem/158126216</a>

Reviewed by Alex Christensen.

This change exposed dispatched(From|To) annotations to the IPCTestingAPI.
This means that tooling can determine whether or not a certain message
is expected receivable directly from the API.

* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_names_header.ProcessName):
(generate_message_names_header):
(generate_message_names_implementation):
* Source/WebKit/Scripts/webkit/model.py:
(generate_global_model):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
(IPC::processLiteral):
(IPC::dispatchedFrom):
(IPC::dispatchedTo):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSIPC::messages):

Canonical link: <a href="https://commits.webkit.org/298710@main">https://commits.webkit.org/298710@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2979ad93b933f2d3a057f324c4645d22d56cb0b1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116347 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36008 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26583 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122404 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66908 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36702 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44596 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88373 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42855 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119296 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29275 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104403 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68805 "Passed tests") | | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/115721 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28357 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22506 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66085 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98657 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22667 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125553 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43241 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32475 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97080 "") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43606 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100609 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96876 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24662 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42164 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20078 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39194 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43128 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48720 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42595 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45935 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44300 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/MessageNames.h:
```diff
@@ -291,6 +302,36 @@ constexpr bool messageIsSync(MessageName name)
+    default:
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+}
```
This doesn't all need to be in a header that's included everywhere.  That'll slow down the build.  Just have function declarations here and the bulk in the cpp.  It's only used for debug builds, so it doesn't need to be constexpr.

gavin-apple replied:
Done.

achristensen07 commented on Source/WebKit/Scripts/webkit/model.py:
```diff
@@ -135,6 +135,6 @@ def generate_global_model(receivers):
             if message.reply_parameters is not None and not message.has_attribute(SYNCHRONOUS_ATTRIBUTE):
                 async_reply_messages.append(Message(name='%s_%sReply' % (receiver.name, message.name), parameters=message.reply_parameters, reply_parameters=[], attributes=None, condition=message.condition))
-    async_reply_receiver = MessageReceiver(name='AsyncReply', superclass='None', attributes=[BUILTIN_ATTRIBUTE], receiver_enabled_by=None, receiver_enabled_by_exception=False, receiver_enabled_by_conjunction=None, receiver_dispatched_from=None, receiver_dispatched_from_exception=None, receiver_dispatched_to=None, receiver_dispatched_to_exception=None, shared_preferences_needs_connection=False, me ...
+    async_reply_receiver = MessageReceiver(name='AsyncReply', superclass='None', attributes=[BUILTIN_ATTRIBUTE], receiver_enabled_by=None, receiver_enabled_by_exception=False, receiver_enabled_by_conjunction=None, receiver_dispatched_from=None, receiver_dispatched_from_exception=True, receiver_dispatched_to=None, receiver_dispatched_to_exception=True, shared_preferences_needs_connection=False, me ...
```
Why are you changing the default of receiver_dispatched_from_exception from None to True?  That seems like a step in the wrong direction.

gavin-apple replied:
It's not a default. We have a pair a "fake/builtin" message receivers (ones that aren't derived from a .in file). One to hold all AsyncReply messages and one to hold a set of IPC builtin functions. They don't (currently) have a defined dispatched(From|To) so I was just trying to represent this via the exceptions in their python MessageReceiver representation. 

Either way, it's not needed for this change so I've reverted and can revisit in a follow-up. It makes sense to try and populate the AsyncReplies using the inverse of the dispatched(From|To) for the outbound message instead of having a permanent exception. However,  that's going to require a bigger change as today they're all just hanging off this one giant global AsyncReply interface.

gavin-apple replied:
Filed: https://bugs.webkit.org/show_bug.cgi?id=297326

achristensen07 reviewed (approved):
It would be good to have the test output some indication of success rather than just no output.

gavin-apple committed:
Expose dispatchedTo and dispatchedFrom annotations to the IPCTestingAPI
https://bugs.webkit.org/show_bug.cgi?id=297269
rdar://problem/158126216

Reviewed by Alex Christensen.

This change exposed dispatched(From|To) annotations to the IPCTestingAPI.
This means that tooling can determine whether or not a certain message
is expected receivable directly from the API.

* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_names_header.ProcessName):
(generate_message_names_header):
(generate_message_names_implementation):
* Source/WebKit/Scripts/webkit/model.py:
(generate_global_model):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
(IPC::processLiteral):
(IPC::dispatchedFrom):
(IPC::dispatchedTo):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSIPC::messages):

Canonical link: https://commits.webkit.org/298710@main
